src/databricks/labs/ucx/hive_metastore/migration_status.py

```diff
@@ -0,0 +1,132 @@
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.errors import NotFound
+
+from databricks.labs.ucx.framework.crawlers import CrawlerBase
+from databricks.labs.ucx.framework.utils import escape_sql_identifier
+from databricks.labs.ucx.hive_metastore import TablesCrawler
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class MigrationStatus:
+    src_schema: str
+    src_table: str
+    dst_catalog: str | None = None
+    dst_schema: str | None = None
+    dst_table: str | None = None
+    update_ts: str | None = None
+
+    def destination(self):
+        return f"{self.dst_catalog}.{self.dst_schema}.{self.dst_table}".lower()
+
+
+@dataclass
+class TableView:
+    catalog: str
+    schema: str
+    name: str
+
+    def key(self):
+        return f"{self.catalog}.{self.schema}.{self.name}".lower()
+
+
+class MigrationIndex:
+    def __init__(self, tables: list[MigrationStatus]):
+        self._index = {(ms.src_schema, ms.src_table): ms for ms in tables}
+
+    def is_migrated(self, schema: str, table: str) -> bool:
+        """Check if a table is migrated."""
+        return self.get(schema, table) is not None
+
+    def get(self, schema: str, table: str) -> MigrationStatus | None:
+        """Get the migration status for a table. If the table is not migrated, return None."""
+        dst = self._index.get((schema.lower(), table.lower()))
+        if not dst or not dst.dst_table:
+            return None
+        return dst
+
+
+class MigrationStatusRefresher(CrawlerBase[MigrationStatus]):
+    def __init__(self, ws: WorkspaceClient, sbe: SqlBackend, schema, table_crawler: TablesCrawler):
+        super().__init__(sbe, "hive_metastore", schema, "migration_status", MigrationStatus)
+        self._ws = ws
+        self._table_crawler = table_crawler
+
+    def snapshot(self) -> Iterable[MigrationStatus]:
+        return self._snapshot(self._try_fetch, self._crawl)
+
+    def index(self) -> MigrationIndex:
+        return MigrationIndex(list(self.snapshot()))
+
+    def get_seen_tables(self) -> dict[str, str]:
+        seen_tables: dict[str, str] = {}
+        for schema in self._iter_schemas():
+            try:
+                tables = self._ws.tables.list(catalog_name=schema.catalog_name, schema_name=schema.name)
+            except NotFound:
+                logger.warning(
+                    f"Schema {schema.catalog_name}.{schema.name} no longer exists. Skipping checking its migration status."
+                )
+                continue
+            for table in tables:
+                if not table.properties:
+                    continue
+                if "upgraded_from" not in table.properties:
+                    continue
+                if not table.full_name:
+                    logger.warning(f"The table {table.name} in {schema.name} has no full name")
+                    continue
+                seen_tables[table.full_name.lower()] = table.properties["upgraded_from"].lower()
```

Comment:
I may be wrong but just wanted to ensure that table.full_name includes catalog and schema. If not, isn't possible that we have same table name exist in two different catalog/schema?